### PR TITLE
[FEATURE] Add examples for localization in plain PHP

### PR DESCRIPTION
**Pull request by linawolf**
references https://github.com/TYPO3-Documentation/TYPO3CMS-Reference-CoreApi/issues/2286

### CODE REVIEW

File: Classes/Controller/Haiku/AjaxController.php [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+namespace T3docs\Examples\Controller\Haiku;
+
+use Psr\Http\Message\ServerRequestInterface;
+use T3docs\Examples\Utility\HaikuSeasonUtility;
+use TYPO3\CMS\Core\Http\PropagateResponseException;
+use TYPO3\CMS\Core\Localization\LanguageService;
+use TYPO3\CMS\Core\Localization\LanguageServiceFactory;
+
+class AjaxController
```
Comment (brotkrueml) [inline Classes/Controller/Haiku/AjaxController.php:11]:
```suggestion
final class AjaxController
```

You know, I like final classes ;-)

File: Classes/Controller/Haiku/AjaxController.php [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    {
+        return $this->languageServiceFactory->createFromSiteLanguage(
+            $request->getAttribute('language')
+            ?? $request->getAttribute('site')->getDefaultLanguage()
+        );
+    }
+
+    public function main(string $content, array $conf, ServerRequestInterface $request): string
+    {
+        $this->conf = $conf;
```
Comment (brotkrueml) [inline Classes/Controller/Haiku/AjaxController.php:31]: Not necessary as not used.

```suggestion
```

File: Classes/Controller/Haiku/AjaxController.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        $this->languageService = $this->getLanguageService($request);
+
+        $parameter = $request->getQueryParams()['tx_examples_haiku']??[];
```
Comment (brotkrueml) [inline Classes/Controller/Haiku/AjaxController.php:34]:
```suggestion
        $parameter = $request->getQueryParams()['tx_examples_haiku'] ?? [];
```

File: Classes/Controller/Haiku/AjaxController.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+    private function translateSeasonAction(string $season)
+    {
+        if (!in_array($season, HaikuSeasonUtility::getSeasons())) {
+            $this->notFoundAction('Season ' . $season . ' not found');
+        }
+        return HaikuSeasonUtility::getSeasonTranslation($season);
+    }
+
+    /**
+     * @throws PropagateResponseException
```
Comment (brotkrueml) [inline Classes/Controller/Haiku/AjaxController.php:75]: This is not true.

File: Classes/Controller/Haiku/AjaxController.php [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+     */
+    private function notFoundAction(string $reason)
+    {
+        throw new \RuntimeException($reason);
```
Comment (brotkrueml) [inline Classes/Controller/Haiku/AjaxController.php:79]: As the exception class is returned in the error JSON, this should not be RuntimeException. Maybe NotFoundException. Or don't return the exception class in JSON. The reason should be enough.

File: Classes/Controller/Haiku/AjaxController.php [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        $this->conf = $conf;
+        $this->languageService = $this->getLanguageService($request);
+
+        $parameter = $request->getQueryParams()['tx_examples_haiku']??[];
+        $action = $parameter['action'] ?? '';
+        try {
+            $result = match ($action) {
+                'getSeasonList' => $this->getTranslatedSeasonListAction(),
+                'translateSeason' => $this->translateSeasonAction($parameter['season'] ?? ''),
+                default => $this->notFoundAction('Action ' . $action . ' not found.'),
```
Comment (brotkrueml) [inline Classes/Controller/Haiku/AjaxController.php:40]:
```suggestion
                default => throw new WhateverException('Action ' . $action . ' not found.'),
```

Throw exception directly.

**Review by brotkrueml — COMMENTED** [review]
Made some comments, haven't tested yet.

But: Why not using a Middleware? For an Ajax request which returns JSON this would be the preferred solution. 

Why would it be the preferred solution? Controllers are widely used for this.

Why introducing HaikuSeasonUtility?

File: Classes/Controller/Haiku/AjaxController.php (hunk 3 above)
Comment (linawolf) [reply Classes/Controller/Haiku/AjaxController.php:34]: Can we adjust the CI so it catches this? It is kind of stupid to always change such things manually?

**Comment by linawolf** [comment]
> Why introducing HaikuSeasonUtility?

To demonstrate how to translate when you have no DI and no request, like for example in the static method of a Utility class

**Comment by linawolf** [comment]
> But: Why not using a Middleware? For an Ajax request which returns JSON this would be the preferred solution.

Why would it be the preferred solution? Controllers are widely used for this. And controllers are also the most likely point where ppl would want to do translations in PHP

File: Classes/Controller/Haiku/AjaxController.php (hunk 3 above)
Comment (brotkrueml) [reply Classes/Controller/Haiku/AjaxController.php:34]: Yup, is on my todo.

**Comment by linawolf** [comment]
Waiting for your implementation as middleware, closing this

**Pull request closed without merging**